WordPress/gutenberg
# Suggest block pattern transformations that are contextual to the currently selected 'simple' blocks (no InnerBlocks)

## Description

Resolves partially: #28736
Part of: #27575

This PR is a spinoff from: https://github.com/WordPress/gutenberg/pull/29890 for suggesting block pattern transformations that are contextual to the currently selected blocks.

The goal of this PR is to handle 'simple' blocks (without InnerBlocks) first, to better manage the complexity and arbitrariness. This is a complex problem where we need to find the right balance between magic(opinionated actions) and APIs. 

## Notes
1. The core blocks that have been updated to include `role:content` in some of their attributes are:
`core/paragraph`, `core/heading`, `core/list`, `core/quote`, `core/pullquote`, `core/preformatted`, `core/verse`, `core/search`.

2. For patterns to use this transformations need to have declared `blockTypes` property, so it now uses only my test patterns. See: https://github.com/WordPress/gutenberg/pull/30471.

## Rationale behind the current decisions
1. I have skipped any check about transforming selected blocks to available block transforms inside patterns, as I think it would be very confusing to have a really different suggested pattern. An example would be to have a paragraph selected and show patterns that do not include paragraph but have transformed the paragraph to a quote block.
2. To show a suggested pattern, **every selected block must have matched with a pattern's block** . This doesn't necessarily mean that the order will be preserved. You can test this in this PR by selecting a paragraph and two headings (in this order) and see the suggested pattern that has the following order: [heading, paragraph, heading]. Also shown at the gif.
3. If a suggested pattern has more blocks than the selected blocks, add them as are from the pattern. Every pattern block that matched with a selected block will be transformed. IMO this is the main reason for doing this and having block patterns in general. We want to inspire/show users alternative designs.

## APIs
1. The block pattern filtering uses the experimental `blockTypes` API introduced here: https://github.com/WordPress/gutenberg/pull/30471
2. A new `__experimentalRole` property is added to block attribute's API. This property is optional and can mark attributes based on their role. Here is utilized by having `content` value and an example [is the `Heading` block](https://github.com/WordPress/gutenberg/pull/30469/files#diff-801846106160a8e52c94f9316006260c2e6438ba5e6ee086b93d9f5f074122bcR14). This is used by for knowing which attributes we need to retain during the pattern transformation, so as the user can keep their block's content and make use of the styling/presentation attributes of the block pattern.

I have many comments in my code trying to explain as good as I can the flows.

## Testing instructions
1. Test with the demo patterns I created in lib/test-block-patterns.php or register one yourselves
2. Create some paragraphs or headings and make different selections (single or multiple) to observe the previews from the selected blocks.
3. Only if all of the selected blocks can be matched in pattern's blocks, it will be shown as an option


## Screenshots 
![simple transforms](https://user-images.githubusercontent.com/16275880/113406871-e75d9b80-93b4-11eb-8b6f-7267e30ad96b.gif)

## TODO in follow ups
- [ ] More tests for some functions + e2e.

## Checklist:
